sql/opt/exec/explain: TestMaximumMemoryUsage failed

sql/opt/exec/explain.TestMaximumMemoryUsage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19295008?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19295008?buildTab=artifacts#/) on master @ [54966d5f661fdf2c038fe88058ab9dde5fbd26bc](https://github.com/cockroachdb/cockroach/commits/54966d5f661fdf2c038fe88058ab9dde5fbd26bc):


```
=== RUN   TestMaximumMemoryUsage
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/3286699bf278cb16f63aa79f134a388f/logTestMaximumMemoryUsage391936990
    test_log_scope.go:76: use -show-logs to present logs inline
    output_test.go:434: 
        	Error Trace:	pkg/sql/opt/exec/explain/output_test.go:434
        	Error:      	"0" is not greater than "5"
        	Test:       	TestMaximumMemoryUsage
        	Messages:   	expected maximum memory usage to be at least 5 MiB, full output:
        	            	
        	            	planning time: 357µs
        	            	execution time: 7.8s
        	            	distribution: full
        	            	vectorized: true
        	            	plan type: custom
        	            	rows decoded from KV: 10,000 (9.9 MiB, 10,000 gRPC calls)
        	            	cumulative time spent in KV: 15.3s
        	            	maximum memory usage: 0 B
        	            	network usage: 4.5 KiB (6 messages)
        	            	regions: test
        	            	sql cpu time: 73ms
        	            	isolation level: serializable
        	            	priority: normal
        	            	quality of service: regular
        	            	
        	            	• group (hash)
        	            	│ group by: bucket
        	            	│
        	            	└── • scan
        	            	      sql nodes: n2, n3
        	            	      kv nodes: n2, n3
        	            	      regions: test
        	            	      actual row count: 10,000
        	            	      KV time: 15.3s
        	            	      KV rows decoded: 10,000
        	            	      KV bytes read: 9.9 MiB
        	            	      KV gRPC calls: 10,000
        	            	      estimated max memory allocated: 4.1 MiB
        	            	      sql cpu time: 65ms
        	            	      missing stats
        	            	      table: t@t_pkey
        	            	      spans: FULL SCAN
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3286699bf278cb16f63aa79f134a388f/logTestMaximumMemoryUsage391936990
--- FAIL: TestMaximumMemoryUsage (12.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMaximumMemoryUsage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-49063

--- a/pkg/sql/opt/exec/explain/BUILD.bazel
+++ b/pkg/sql/opt/exec/explain/BUILD.bazel
@@ -89,6 +89,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/treeprinter",
         "@com_github_cockroachdb_datadriven//:datadriven",
+        "@com_github_cockroachdb_errors//:errors",
         "@com_github_stretchr_testify//assert",
         "@com_github_stretchr_testify//require",
         "@in_gopkg_yaml_v2//:yaml_v2",

--- a/pkg/sql/opt/exec/explain/output_test.go
+++ b/pkg/sql/opt/exec/explain/output_test.go
@@ -30,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/datadriven"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	yaml "gopkg.in/yaml.v2"
@@ -417,19 +418,29 @@ func TestMaximumMemoryUsage(t *testing.T) {
 		return err
 	})
 
-	rows := db.QueryStr(t, "EXPLAIN ANALYZE SELECT max(v) FROM t GROUP BY bucket;")
-	var output strings.Builder
-	maxMemoryRE := regexp.MustCompile(`maximum memory usage: ([\d\.]+) MiB`)
-	var maxMemoryUsage float64
-	for _, row := range rows {
-		output.WriteString(row[0])
-		output.WriteString("\n")
-		s := strings.TrimSpace(row[0])
-		if matches := maxMemoryRE.FindStringSubmatch(s); len(matches) > 0 {
-			var err error
-			maxMemoryUsage, err = strconv.ParseFloat(matches[1], 64)
-			require.NoError(t, err)
+	// In rare cases (due to metamorphic randomization) we might drop the
+	// ComponentStats proto that powers "maximum memory usage" stat from the
+	// trace, so we add a retry loop around this.
+	testutils.SucceedsSoon(t, func() error {
+		rows := db.QueryStr(t, "EXPLAIN ANALYZE SELECT max(v) FROM t GROUP BY bucket;")
+		var output strings.Builder
+		maxMemoryRE := regexp.MustCompile(`maximum memory usage: ([\d\.]+) MiB`)
+		var maxMemoryUsage float64
+		for _, row := range rows {
+			output.WriteString(row[0])
+			output.WriteString("\n")
+			s := strings.TrimSpace(row[0])
+			if matches := maxMemoryRE.FindStringSubmatch(s); len(matches) > 0 {
+				var err error
+				maxMemoryUsage, err = strconv.ParseFloat(matches[1], 64)
+				if err != nil {
+					return err
+				}
+			}
 		}
-	}
-	require.Greaterf(t, maxMemoryUsage, 5.0, "expected maximum memory usage to be at least 5 MiB, full output:\n\n%s", output.String())
+		if maxMemoryUsage < 5.0 {
+			return errors.Newf("expected maximum memory usage to be at least 5 MiB, full output:\n\n%s", output.String())
+		}
+		return nil
+	})
 }

